src/communication/distributed_context.hpp

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include <string>
+
+#include <spike.hpp>
+#include <communication/gathered_vector.hpp>
+#include <util/pp_util.hpp>
+
+#if defined(ARB_HAVE_MPI)
+#   include "mpi_context.hpp"
+#endif
+#include "local_context.hpp"
+
+
+namespace arb {
+
+#define ARB_PUBLIC_COLLECTIVES_(T) \
```

Comment:
This way it is very simple to
  * use a context: because it's interface is type erased.
  * define a new context implementation: see the `mpi_context` and `local_context` implementations, which are very simple and clean, with no need to inherit from or know about the `distributed_context` type.
  * I like this specific type erasure because the user doesn't need to mess around with pointers to base class.